Pagination for large libraries

GAntoine opened the pull request:
There's considerable slowdown (lag) in the app when the number of roms in a library hits 500+. This PR attempts to mitigate that issue in a couple ways.

### What changed?

* Introduce cursor-based pagination with [fastapi_pagination](https://uriyyo-fastapi-pagination.netlify.app/)
  * Gallery view holds next cursor value and requests further entries with it
  * ~~Add [axios-cache-interceptor](https://axios-cache-interceptor.js.org/) to cached cursored rom list requests~~
* Set `timestamp` on cover url when cover is updated
  * This allows for cache-busting images but only when required
* Move search to backend and use rom list endpoint with `ilike` filter
  * Add [lodash](https://lodash.com/) to debounce search input
* Fix accessing `file_name` when updating rom, but "Rename rom" not checked

### What's left?

- [x] Get [vue-virtual-scroller](https://github.com/Akryum/vue-virtual-scroller) working, at least for the list view.
- [x] Use `onScroll` instead of needing a dedicated button to load more

### How do I test this?

Either already have a very large library, or generate a ton of empty ZIP files and run a scan to load them in the DB.

```sh
for i in {0001..1000}; do touch "file_${i}.zip"; done
```

Closes #280 
Closes #89

zurdi15 commented:
I had a lot of front code refactor in my local, so I am going to push them to master to let you resolve any kind of conflict that will appear as I see you changed some things too. I am sorry for the inconveniences you are going to find with this.

GAntoine commented:
> so I am going to push them to master to let you resolve any kind of conflict that will appear as I see you changed some things too

yeah that's fine, my work should be targeted enough that i can just splice it in. might need a new branch/PR but we'll see.

zurdi15 commented on backend/endpoints/rom.py:
```diff
@@ -10,47 +13,115 @@
 router = APIRouter()
 
 
+class RomSchema(BaseModel):
```
Is there a way to `merge` somehow this `RomSchema` with the `Rom` model? I guess as all the `Rom` attributes are `Columns` it can't be translated directly to python types, but I was just wondering if we can just have a single place where to put the models schamas and extend from there.

GAntoine commented on backend/endpoints/rom.py:
```diff
@@ -10,47 +13,115 @@
+    path_screenshots: list
+
+    class Config(BaseModel.Config):
+        orm_mode = True
```
At some point I'd like to move these model definitions into their own files, once we have more of them

zurdi15 replied:
I was asking the same, didn't saw this comment. I agree with that, I guess for now we can leave it here and move it when it grows a little bit. For now it is manageable

GAntoine commented on backend/handler/db_handler.py:
```diff
@@ -75,56 +69,51 @@ def add_rom(self, rom: Rom) -> None:
         try:
-            with self.session.begin() as s:
-                return s.scalars(select(Rom).filter_by(p_slug=p_slug).order_by(Rom.file_name.asc())).all()
+            return select(Rom).filter_by(p_slug=p_slug).order_by(Rom.file_name.asc())
```
Only change here is removing the scalar

zurdi15 replied:
Yes totally agree. To be honest I was learning SQAlchemy while developing RomM, so I see now a lot of imporvements with it.

zurdi15 commented:
Let me test the branch before merging, but everything seems really OK for me.

GAntoine commented:
> Let me test the branch before merging

Yes please! The most important things to test are pagination, search in a library, updating and deleting ROMs, and the 3 different views (big, small, list).

zurdi15 commented:
First of all I must say that the load on scroll works perfectly fine, but I tested everything a little bit and I found some problems:
 - Pagination doesn't work for me in the big and small view, just in the list view
 - Editing the rom broke. Now the edit dialog doesn't show the rom info and can't edit it
 - Delete rom works fine, but I notice now that after delete/scan something the you need to reload the page to show the changes, I think is a good thing to reload the gallery itself after delete or scan.
 - Search in library works fine, but I realized that when searching with the filter, the pagination doesn't work, it will only show the first 100 entries it finds.
 
 I'll keep testing it anyway.

GAntoine commented:
> Pagination doesn't work for me in the big and small view, just in the list view

Scrolling wasn't hooked up correctly, fixed now

> Now the edit dialog doesn't show the rom info and can't edit it

Yep forgot to replace a `.data.data`, fixed

> after delete/scan something the you need to reload the page to show the changes

I removed `axios-cache-interceptor` which was causing the issue, we can revisit smarter caching at a later date

> when searching with the filter, the pagination doesn't work

Yeah I never build that, added it and seems to work, though there is 1 bug I need to fix

zurdi15 commented:
Wow you are fast. I will test it asap and merge :)

GAntoine commented:
@zurdi15 I had a though: instead of using an infinite scroll, what it we did some simple [pagination](https://vuetifyjs.com/en/components/paginations/)? It would limit the slowdown of having to render hundreds of rows/cards, and we could do it in chunks of 54 or so.

GAntoine commented:
Good catch, both issues should be fixed 👍🏼
